src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs

```diff
@@ -39,6 +37,18 @@ internal SeparatedWithManyChildren(InternalSyntax.SyntaxList green, SyntaxNode?
 
                 return _children[i >> 1].Value;
             }
+
+            internal override int GetChildPosition(int index)
+            {
+                // If the previous sibling (ignoring separator) is not cached, but the next sibling
+                // (ignoring separator) is cached, use the next sibling to determine position.
+                int valueIndex = (index & 1) != 0 ? index - 1 : index;
+                bool useNextNotPrevious = valueIndex > 1
+                    && _children[(valueIndex - 2) >> 1].Value is null
+                    && (valueIndex >= Green.SlotCount - 2 || _children[(valueIndex + 2) >> 1].Value is { });
+
+                return GetChildPosition(index, useNextNotPrevious);
```

Comment:
We decide whether to look at previous siblings or following siblings in `GetChildPosition()` by looking at the nearest siblings. If the nearest previous sibling is not included in the `_children` cache, but the nearest following sibling is cached, we'll look at the following siblings. Otherwise, we'll use the existing behavior of looking at the previous siblings. To get the nearest siblings, we ignore separators though, because those are not represented in the `_children` cache.